src/compound_model/CompoundModelFactory.py

```diff
@@ -0,0 +1,403 @@
+            replace (bool, optional): True if samples in the population should be replaced, if false samples are appended to
+                the population instead. Defaults to False.
+
+        Raises:
+            AssertionError: When `num_samples` or `num_attrs` are not valid values.
+        """
+
+        # Check input
+        assert num_samples > 0
+        assert num_attrs >= -1
+        assert num_attrs <= len(self._controller.get_attributes())
+
+        pop_samples = self._population.num_samples()
+
+        # Check early stopping criteria
+        if (not replace) and pop_samples >= num_samples:
+            print(
+                f"The population already consists of {pop_samples} ({num_samples} was the goal) samples."
+            )
+            return
+
+        # Set number of samples to generate
+        samples_to_gen = num_samples if replace else num_samples - pop_samples
+
+        # Set sample mode, see sample_random_input() docs for more info
+        attr_sample_mode = "random" if num_attrs == -1 else num_attrs
+        random_input = self._controller.sample_random_input(
+            samples_to_gen, attr_sample_mode
+        )
+
+        # Parse input
+        parsed_input = self._controller.parse_native_input(random_input)
+
+        # Get latent codes
+        latent_codes = self._generator.random_latent_code(samples_to_gen)
+
+        # Generate images
+        uris, manipulated_latent_codes = self._controller.generate_native(
+            latent_codes, parsed_input
+        )
+
+        # Add images to population
+        self._population.add_all(
+            manipulated_latent_codes,
+            latent_codes,
+            uris,
+            [1] * len(uris),
+            append=(not replace),
+            **random_input,
+        )
+
+        # Clear compound metrics
+        self._clear_compound_metrics(self._population.get_name())
+
+    def _clear_compound_metrics(self, population_name: str) -> None:
+        # Clear compound metrics from manager
+        for filter in self._filters:
+            cmm = self._cmms[filter.get_name()]
+            if population_name == cmm.get_population().get_name():
+                cmm.clear_compound_metrics()
+
+        # Clear locally stored compound metrics from population
+        CompoundMetricManager.clear_local_compound_metrics(population_name)
+
+    def evaluate_compound_models(
+        self,
+        filters: Union[Type[Filter], list[Type[Filter]]] = None,
+        compound_metrics: Union[list[str], str] = None,
+        recalculate_metrics: bool = False,
+        **parameters: Any,
+    ) -> None:
+        """
+        Evaluates a selection or all of the compound models on a selection of compound metrics.
+        Result are saved to the managers storage.
+
+        Args:
+            filters (Union[Type[Filter], list[Type[Filter]]], optional): Selection of filters to use, must be subset
```

Comment:
instead of "Union[Type[Filter], list[Type[Filter]]]" you can put "Type[Filter] | list[Type[Filter]]" if you want to, same goes for all other Union things in this file